src/freertos_drivers/esp32/Esp32Gpio.hxx

```diff
@@ -0,0 +1,467 @@
+/** \copyright
+ * Copyright (c) 2020, Mike Dunston
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are  permitted provided that the following conditions are met:
+ *
+ *  - Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ *  - Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * \file Esp32Gpio.hxx
+ *
+ * Helper declarations for using GPIO pins via the ESP-IDF APIs.
+ *
+ * @author Mike Dunston
+ * @date 27 March 2020
+ */
+
+#ifndef _DRIVERS_ESP32GPIO_HXX_
+#define _DRIVERS_ESP32GPIO_HXX_
+
+#include "freertos_drivers/arduino/GpioWrapper.hxx"
+#include "os/Gpio.hxx"
+#include "utils/logging.h"
+#include "utils/macros.h"
+
+#include <driver/gpio.h>
+
+/// Helper macro to test if a pin has been configured for output.
+///
+/// This is necessary since ESP-IDF does not expose gpio_get_direction(pin).
+#define IS_GPIO_OUTPUT(pin) (GPIO_IS_VALID_OUTPUT_GPIO(pin) &&                 \
+                             (pin < 32) ? GPIO.enable & BIT(pin & 31) :        \
+                                          GPIO.enable1.data & BIT(pin & 31))
+
+template <class Defs, bool SAFE_VALUE, bool INVERT> struct GpioOutputPin;
+template <class Defs, bool PUEN, bool PDEN> struct GpioInputPin;
+
+/// Defines a GPIO output pin. Writes to this structure will change the output
+/// level of the pin. Reads will return the pin's current level.
+///
+/// The pin is set to output at initialization time, with the level defined by
+/// `SAFE_VALUE'.
+///
+/// Do not use this class directly. Use @ref GPIO_PIN instead.
+template <gpio_num_t PIN_NUM, bool INVERTED = false>
```

Comment:
There is now an InvertedGpio template. Directly supporting inversion in the base template is no longer necessary. This could decrease code complexity in this file.

Reply:
I'll take a look at that.